Add Faucet contract

sebastinez opened the pull request:
This PR implements a faucet contract for the Rinkeby Testnet.

Feedback is welcome

cloudhead commented on bin/deploy:
```diff
@@ -0,0 +1,8 @@
+export ETH_GAS=2000000
+export GAS_PRICE=1gwei
+
+export DAPP_BUILD_OPTIMIZE_RUNS=1
```
For a contract like this I would optimize towards many runs, so I'd set this to `200`

cloudhead commented on src/FaucetV1.sol:
```diff
@@ -0,0 +1,103 @@
+    /// @return Bool if requester is allowed to create more withdrawals
+    function checkConsecutiveWithdrawals(address _requester) private view returns (bool) {
+        return consecutiveWithdrawals[_requester] < maxConsecutiveWithdrawals;
+    }
```
Can we avoid checking for consecutive withdrawals completely, and just check for the amount per time unit? I'm not sure we gain anything from limiting the number of times someone withdraws, since they pay for the gas anyway.

sebastinez replied:
The current version has two ways of restricting withdrawals.
- One is per time via a quadratic function, that timelocks the withdrawal function for a specific time.
- The other is to limit consecutive withdrawals made in a period of 24 hours.

The reason behind it is that without, one could make small amounts every few seconds and like that slowly drain the faucet..

cloudhead replied:
Hmm, I get it, but I think it would be simpler and cheaper if we prevented that with a single time-based cap, no? Eg. `min(timeSinceLastWithdrawal, 1 day) * multiplier`

sebastinez replied:
Hmm I think we could do a single time based cap, and eventually raise the minimum amount to wait, to for example 30 minutes?

I'm not quite sure what you meant with the multiplier in your example.

sebastinez replied:
Hmm I think we could do a single time based cap, to save on deployment costs.

Since the user will be paying for the gas, we could just remove the consecutive withdrawals and leave the time based cap as it is currently.
Eventually the user will be spending more Rinkeby ETH on gas fees, for obtaining less to nothing on Rinkeby RAD.

cloudhead replied:
Yeah, that's what I'm thinking!

Commit:
Add Faucet Contract

cloudhead reviewed (approved):
Nice work!

sebastinez commented:
Just FYI, I removed the paused functions, since I think it's easier to reduce the `maxWithdrawAmount` to zero and to pause the contract like this.
I added also a `recoverFunds` function, since while testing it happended quite a lot to deploy, have to redeploy and not able to get the funds out.. :sweat_smile:
